src/events.rs

```diff
@@ -372,29 +372,112 @@ impl<'js> EventEmitter<'js> {
     }
 }
 
-//TODO implement, abort(reason)
 #[rquickjs::class]
 #[derive(rquickjs::class::Trace)]
-struct AbortController {}
+pub struct AbortController<'js> {
+    signal: Class<'js, AbortSignal<'js>>,
+}
 
 #[rquickjs::methods]
-impl AbortController {
+impl<'js> AbortController<'js> {
     #[qjs(constructor)]
-    fn new() -> Self {
-        Self {}
+    pub fn new(ctx: Ctx<'js>) -> Result<Self> {
+        let abort_controller = Self {
+            signal: Class::instance(
+                ctx,
+                AbortSignal {
+                    aborted: false,
+                    reason: None,
+                },
+            )?,
+        };
+        Ok(abort_controller)
+    }
+
+    #[qjs(get)]
+    pub fn signal(&self) -> Class<'js, AbortSignal<'js>> {
+        self.signal.clone()
+    }
+
+    pub fn abort(
+        this: This<Class<'js, Self>>,
+        reason: Opt<Value<'js>>,
+    ) -> Result<Class<'js, Self>> {
+        this.0.borrow_mut().signal.borrow_mut().set_aborted(true);
+        if reason.0.is_some() {
+            this.0.borrow_mut().signal.borrow_mut().set_reason(reason)?;
+        } else {
+            // TODO store DOMException as reason
+            this.0.borrow_mut().signal.borrow_mut().set_reason(reason)?;
+        }
+
+        Ok(this.0)
     }
 }
 
 //TODO implement static methods abort() and timeout(miliseconds)
 #[rquickjs::class]
-#[derive(rquickjs::class::Trace)]
-struct AbortSignal {}
+pub struct AbortSignal<'js> {
+    aborted: bool,
+    reason: Option<Value<'js>>,
+}
+
+impl<'js> Trace<'js> for AbortSignal<'js> {
+    fn trace<'a>(&self, tracer: Tracer<'a, 'js>) {
+        if let Some(reason) = &self.reason {
+            tracer.mark(reason);
+        }
+    }
+}
 
 #[rquickjs::methods]
-impl AbortSignal {
+impl<'js> AbortSignal<'js> {
     #[qjs(constructor)]
-    fn new() -> Self {
-        Self {}
+    pub fn new() -> Self {
+        Self {
+            aborted: false,
+            reason: None,
+        }
+    }
+
+    #[qjs(get)]
+    pub fn aborted(&self) -> bool {
+        self.aborted
+    }
+
+    #[qjs(get)]
+    pub fn reason(&self) -> Option<Value<'js>> {
+        self.reason.clone()
+    }
+
+
+    #[qjs(skip)]
+    pub fn set_reason(&mut self, reason: Opt<Value<'js>>) -> Result<Option<Value<'js>>> {
+        if let Some(new_reason) = reason.0 {
+            Ok(self.reason.replace(new_reason))
+        } else {
+            let old_reason = self.reason.take();
+            Ok(old_reason)
+        }
+    }
+
+    #[qjs(skip)]
+    pub fn set_aborted(&mut self, is_aborted: bool) {
+        self.aborted = is_aborted;
+    }
+
+    #[qjs(static)]
+    pub fn abort(reason: Opt<Value<'js>>) -> AbortSignal {
+        AbortSignal {
+            aborted: true,
+            reason: reason.0,
+        }
+    }
+
+    // TODO: Returns an AbortSignal that will automatically abort after a specified time.
+    #[qjs(static)]
+    pub fn timeout() -> AbortSignal<'js> {
+        todo!()
```

Comment:
Here you can simply use `ctx.spawn_exit` to spawn a task and then set abort after the specified time. First create a new abort signal and convert it to a Value: `AbortSignal{...}.into_js()`. Then clone it and return one copy while pass the other copy to the async task which then sets the abort value after said time. Something like:
```rust
 pub fn timeout<'js>(ctx: Ctx<'js>, mills: size) -> Result<Class<'js,Self>> {
        let signal = AbortSignal {
            aborted: true,
            reason: reason.0,
        }
        
        let signal_instance =  Class::instance(ctx,signal);
        let signal_instance2 = signal_instance.clone()
        
        ctx.spawn_exit(async move {
           tokio::time::sleep(mills).await;
           let borrow = signal_instance.borrow();
           borrow.aborted = true;
           borrow.reason = "Timeout" //check spec
        })?;
        Ok(signal_instance2);
    }

Reply:
Thanks! That's really neat 😮 One question about the spec - it mentions putting `DOMException` objects as the `reason` in different cases. I'm not sure if I can I just use the quickjs `Exception` struct here - it looks like `DOMException` is a specific [interface type](https://webidl.spec.whatwg.org/#idl-DOMException). I could include making that in this PR if necessary

Reply:
Mark it as a todo and we'll address it later :) DOMEx feels a bit out of scope for this PR

Reply:
Wasn't sure the best way to test this, doesn't seem practical with `jest`